File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        res=[]
        n=len(nums)
        nums.sort()
        for i in range(n):
            if i>0 and nums[i]==nums[i-1]:
                continue
            l=i+1
            r=n-1
            while l<r:
                total=nums[i]+nums[l]+nums[r]
                if total==0:
                    res.append([nums[i],nums[l],nums[r]])
                    while l<r and nums[l]==nums[l+1]:
                        l+=1
                    while r>l and nums[r]==nums[r-1]:
                        r-=1
                    l+=1
                    r-=1
                elif total<0:
                    l+=1
                else:
                    r-=1
        return res

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
       ArrayList <List<Integer>> res=new ArrayList<>();
       Arrays.sort(nums);
       for (int i=0;i<nums.length;i++){
        if (i>0 && nums[i-1]==nums[i]){
            continue;
        }
        int left=i+1;
        int right=nums.length-1;
        while (left<right){
            int total=nums[left]+nums[right]+nums[i];
            if (total==0){
                res.add(Arrays.asList(nums[i],nums[left],nums[right]));
            
                while (left<right && nums[left]==nums[left+1]){
                    left+=1;
                }
                while (left<right && nums[right]==nums[right-1]){
                    right-=1;
                }
                left+=1;
                right-=1;
            }
            else if (total<0){
                left++;
            }
            else{
                right--;
            }
        }
       } 
       return res ;
    }
}